src/Dom/ResourceHintManager.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+namespace AmpProject\Dom;
+
+use AmpProject\Attribute;
+use AmpProject\RequestDestination;
+use AmpProject\Tag;
+use DOMNode;
+
+/**
+ * Resource hint manager class that is used to manage the <link> tags within a document's <head> to give the browser
+ * hints about how to prioritize resources.
+ *
+ * @package ampproject/amp-toolbox
+ */
+final class ResourceHintManager
+{
+
+    /**
+     * Document to manage the resource hints for.
+     *
+     * @var Document
+     */
+    private $document;
+
+    /**
+     * Reference node to attach the resource hint to.
+     *
+     * @var DOMNode|null
+     */
+    private $referenceNode;
+
+    /**
+     * ResourceHintManager constructor.
+     *
+     * @param Document $document
+     */
+    public function __construct(Document $document)
+    {
+        $this->document = $document;
+    }
+
+    /**
+     * Add a preconnect resource hint.
+     *
+     * @param string $href        URL to link to.
+     * @param bool   $crossorigin Whether the link is crossorigin.
+     */
+    public function addPreconnect($href, $crossorigin = true)
+    {
+        $this->addLinkTag(
+            Attribute::REL_PRECONNECT,
+            $href,
+            $crossorigin ? [ Attribute::CROSSORIGIN => null ] : []
+        );
+
+        // Use dns-prefetch as fallback for browser that don't support preconnect.
+        // See https://web.dev/preconnect-and-dns-prefetch/#resolve-domain-name-early-with-reldns-prefetch
+        $this->addLinkTag(Attribute::REL_DNS_PREFETCH, $href);
+    }
+
+    /**
+     * Add a preconnect resource hint.
+     *
+     * @param string      $href  URL to link to.
+     * @param string      $type  Optional. Type of the resource. Defaults to type 'image'.
+     * @param string|null $media Optional. Media query to add to the preload. Defaults to none.
+     */
+    public function addPreload($href, $type = RequestDestination::IMAGE, $media = null)
+    {
+        // TODO: Should we enforce a valid $type here?
+
+        $attributes = [ Attribute::AS_ => $type ];
+        if (!empty($media)) {
+            $attributes[Attribute::MEDIA] = $media;
+        }
+
+        $this->addLinkTag(Attribute::REL_PRELOAD, $href, $attributes);
+    }
+
+    /**
+     * Add a link tag.
+     *
+     * @param string   $rel        A 'rel' string.
+     * @param string   $href       URL to link to.
+     * @param string[] $attributes Associative array of attributes and their values.
+     */
+    public function addLinkTag($rel, $href, $attributes = [])
```

Comment:
This should perhaps be `private` as otherwise it a user could add arbitrary link relations via this manager, including `stylesheet`, `glossary`, etc.

This being the case, should there be additional method for `modulepreload`? Should there be one for `dns-prefetch`? There seems to be value in adding `dns-prefetch` _without_ also adding `preconnect` for non-critical resources per [MDN](https://developer.mozilla.org/en-US/docs/Web/Performance/dns-prefetch#:~:text=if%20a%20page%20needs%20to%20make%20connections%20to%20many%20third-party%20domains):

<blockquote>
<p><strong>Note</strong>: If a page needs to make connections to many third-party domains, preconnecting them all is counterproductive. The <code>preconnect</code> hint is best used for only the most critical connections. For the others, just use <code>&lt;link rel="dns-prefetch"&gt;</code> to save time on the first step — the DNS lookup.</p>
</blockquote>

Reply:
Additional methods would be to add `prefetch` and `prerender` links.